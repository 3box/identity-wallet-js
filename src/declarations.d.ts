// not sure why this is needed
declare module 'ipfs' {
  export default any
}

declare module 'multihashes' {
  export function encode(digest: Buffer, code: number | string, length?: number): Buffer
}
