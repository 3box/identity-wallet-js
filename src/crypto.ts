import { randomBytes } from '@stablelib/random'
export { randomBytes }
import {
  createJWE,
  decryptJWE,
  JWE,
  Encrypter,
  Decrypter,
  x25519Encrypter,
  x25519Decrypter,
} from 'did-jwt'
import { prepareCleartext, decodeCleartext } from 'dag-jose-utils'

interface EncryptOpts {
  publicKey?: Uint8Array
  encrypter?: Encrypter
  kid?: string
}

function isU8a(b: Uint8Array | undefined): b is Uint8Array {
  return b instanceof Uint8Array
}

export async function asymEncryptJWE(
  cleartext: Record<string, any>,
  { publicKey, encrypter, kid }: EncryptOpts
): Promise<JWE> {
  if (!encrypter) {
    if (!isU8a(publicKey)) throw new Error('publicKey or encrypter has to be defined')
    encrypter = x25519Encrypter(publicKey, kid)
  }
  return createJWE(prepareCleartext(cleartext), [encrypter])
}

interface DecryptOpts {
  secretKey?: Uint8Array
  decrypter?: Decrypter
}

export async function asymDecryptJWE(
  jwe: JWE,
  { secretKey, decrypter }: DecryptOpts
): Promise<Record<string, any>> {
  if (!decrypter) {
    if (!isU8a(secretKey)) throw new Error('secretKey or decrypter has to be defined')
    decrypter = x25519Decrypter(secretKey)
  }
  return decodeCleartext(await decryptJWE(jwe, decrypter))
}

interface Recipient {
  header: Record<string, string>
}

export function parseJWEKids(jwe: JWE): Array<string> {
  return (jwe.recipients || []).reduce((kids: Array<string>, recipient: Recipient) => {
    if (recipient.header?.kid) kids.push(recipient.header.kid.split('#')[1])
    return kids
  }, [])
}
