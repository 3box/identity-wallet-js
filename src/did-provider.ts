import { createJWS } from 'did-jwt'
import {
  HandlerMethods,
  RequestHandler,
  RPCConnection,
  RPCError,
  RPCRequest,
  RPCResponse,
  createHandler,
} from 'rpc-utils'

import Keyring from './keyring'
import { ThreeIDX } from './three-idx'
import Permissions from './permissions'
import { toStableObject } from './utils'

type Origin = string | null | undefined

export type Context = {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  origin: Origin
}

interface CreateJWSParams {
  payload: Record<string, any>
  protected?: Record<string, any>
  did: string
}

interface AuthParams {
  paths: Array<string>
}

export const didMethods: HandlerMethods<Context> = {
  did_authenticate: async ({ permissions, threeIdx, origin }, params: AuthParams) => {
    const paths = await permissions.request(origin, params.paths || [])
    // paths should be an array if permission granted
    // may be a subset or requested paths or empty array
    if (paths === null) throw new RPCError(4001, 'User Rejected Request')
    return { did: threeIdx.DID, paths }
  },
  did_createJWS: async ({ permissions, keyring, threeIdx, origin }, params: CreateJWSParams) => {
    if (!permissions.has(origin)) {
      throw new RPCError(4100, 'Unauthorized')
    }
    // TODO - if the requesting DID is our management key
    // (did:key) we should request explicit permission.
    const keyName = threeIdx.parseKeyName(params.did)
    const kid = await threeIdx.encodeKidWithVersion(keyName)
    const signer = keyring.getSigner(keyName)
    const header = toStableObject(Object.assign(params.protected || {}, { kid }))
    const jws = await createJWS(toStableObject(params.payload), signer, header)
    return { jws }
  },
}

export interface ProviderConfig {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  forcedOrigin?: string
}

export class DidProvider implements RPCConnection {
  protected _handle: RequestHandler

  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
    const handler = createHandler<Context>(didMethods)
    this._handle = (origin: string, msg: RPCRequest) => {
      return handler(
        {
          origin: forcedOrigin || origin,
          permissions,
          threeIdx,
          keyring,
        },
        msg
      )
    }
  }

  public get isDidProvider(): boolean {
    return true
  }

  public async send(msg: RPCRequest, origin?: Origin): Promise<RPCResponse | null> {
    return await this._handle(origin, msg)
  }
}
