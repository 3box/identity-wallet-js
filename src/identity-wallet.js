import EventEmitter from 'events'
import Keyring from './keyring'
import didJWT from 'did-jwt'
import { sha256Multihash, pad, unpad } from './utils'


class IdentityWallet {

  /**
   * Creates an instance of IdentityWallet
   *
   * @param     {Object}    config                  The configuration to be used
   * @param     {String}    config.seed             The seed of the identity, 32 hex string
   * @param     {String}    config.authSecret       The authSecret to use, 32 hex string
   * @param     {String}    config.ethereumAddress  The ethereumAddress of the identity
   * @return    {this}                            An IdentityWallet instance
   */
  constructor (config = {}) {
    this.events = new EventEmitter()
    if (config.seed) {
      this._keyring = new Keyring(config.seed)
    }
    if (config.authSecret) {
      if (!config.ethereumAddress) throw new Error('Ethereum address needs to be defined when authSecret given')
      this._authSecret = config.authSecret
      this._ethereumAddress = config.ethereumAddress
    }
  }

  async getAddress () {
    return this._ethereumAddress
  }

  _initKeyring (authData) {
    let seed
    if (authData) {
      authData.find(({ ciphertext, nonce }) => {
        const key = Keyring.hexToUint8Array(this._authSecret)
        seed = Keyring.symDecryptBase(ciphertext, key, nonce)
        return Boolean(seed)
      })
      if (!seed) throw new Error('No valid auth-secret for this identity')
    } else {
      // no authData available so we create a new identity
      seed = '0x' + Buffer.from(Keyring.naclRandom(32)).toString('hex')
      this.addAuthMethod(this._authSecret, seed)
    }
    this._keyring = new Keyring(seed)
  }

  /**
   * Authenticate to given spaces
   *
   * @param     {Array<String>}     spaces          The desired spaces
   * @param     {Object}    opts                    Optional parameters
   * @param     {Array<Object>}     opts.authData   The authData for this identity
   * @return    {Object}                            The public keys for the requested spaces of this identity
   */
  async authenticate (spaces = [], { authData } = {}) {
    if (!this._keyring) this._initKeyring(authData)
    const result = {
      main: this._keyring.getPublicKeys(),
      spaces: {}
    }
    if (this._ethereumAddress) result.main.managementKey = this._ethereumAddress
    spaces.map(space => {
      result.spaces[space] = this._keyring.getPublicKeys({ space, uncompressed: true })
    })
    return result
  }

  /**
   * Check if authenticated to given spaces
   *
   * @param     {Array<String>}     spaces          The desired spaces
   * @return    {Boolean}                           True if authenticated
   */
  async isAuthenticated (spaces = []) {
    return Boolean(this._keyring)
  }

  /**
   * Add a new authentication method for this identity
   *
   * @param     {String}    authSecret              A 32 byte hex string used as authentication secret
   */
  async addAuthMethod (authSecret, seed) {
    if (!seed && !this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const message = seed || this._keyring._seed
    const key = Keyring.hexToUint8Array(authSecret)
    const encAuthData = Keyring.symEncryptBase(message, key)

    this.events.emit('new-auth-method', encAuthData)
  }

  /**
   * Sign a verifiable credential. The format of the credential is [did-jwt](https://github.com/uport-project/did-jwt).
   *
   * @param     {Object}    payload                 The payload of the claim
   * @param     {Object}    opts                    Optional parameters
   * @param     {String}    opts.DID                The DID used as the issuer of this claim
   * @param     {String}    opts.space              The space used to sign the claim
   * @param     {String}    opts.expiresIn          Set an expiry date for the claim as unix timestamp
   * @return    {String}                            The signed claim encoded as a JWT
   */
  async signClaim (payload, { DID, space, expiresIn } = {}) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const issuer = DID
    const settings = {
      signer: this._keyring.getJWTSigner(space),
      issuer,
      expiresIn
    }
    return didJWT.createJWT(payload, settings)
  }

  /**
   * Encrypt a message
   *
   * @param     {String}    message                 The message to be encrypted
   * @param     {String}    space                   The space used for encryption
   * @param     {Object}    opts                    Optional parameters
   * @param     {String}    opts.nonce              The nonce used to encrypt the message
   * @param     {String}    opts.blockSize          The blockSize used for padding (default 24)
   * @return    {Object}                            The encrypted object (ciphertext and nonce)
   */
  async encrypt (message, space, { nonce, blockSize } = {}) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const paddedMsg = pad(message, blockSize)
    return this._keyring.symEncrypt(paddedMsg, { space, nonce })
  }

  /**
   * Decrypt a message
   *
   * @param     {Object}    encryptedObject         The encrypted object (ciphertext and nonce)
   * @param     {String}    space                   The space used for encryption
   * @return    {String}                            The decrypted message
   */
  async decrypt (encObj, space) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const paddedMsg = this._keyring.symDecrypt(encObj.ciphertext, encObj.nonce, { space })
    if (!paddedMsg) throw new Error('IdentityWallet: Could not decrypt message')
    return unpad(paddedMsg)
  }

  async hashDBKey(key, space) {
    const salt = this._keyring.getDBSalt(space)
    return sha256Multihash(salt + key)
  }
}


module.exports = IdentityWallet
