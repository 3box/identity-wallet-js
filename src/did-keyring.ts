import nacl, { BoxKeyPair } from 'tweetnacl'
import naclutil from 'tweetnacl-util'
import { HDNode } from '@ethersproject/hdnode'
import { Wallet } from '@ethersproject/wallet'
import { SimpleSigner } from 'did-jwt'
import { sha256 } from 'js-sha256'
import { ec as EC } from 'elliptic'

import { naclRandom } from './crypto'

const ec = new EC('secp256k1')

const BASE_PATH = "m/51073068'/0'"
const ROOT_STORE_PATH = "0'/0'/0'/0'/0'/0'/0'/0'"

const AUTH_PATH_WALLET = BASE_PATH + '/' + ROOT_STORE_PATH + '/0'
const AUTH_PATH_ENCRYPTION = BASE_PATH + '/' + ROOT_STORE_PATH + '/3'

const ensure0x = (str: string): string => {
  return (str.startsWith('0x') ? '' : '0x') + str
}

interface KeysSet {
  signingKey: HDNode
  asymEncryptionKey: BoxKeyPair
  symEncryptionKey: Uint8Array
  managementKey: HDNode
}

export class Keyring {
  public static create() {
    const seed = '0x' + Buffer.from(naclRandom(32)).toString('hex')
    return new Keyring(seed)
  }

  private _seed: string
  private _baseNode: HDNode
  private _keys: KeysSet

  constructor(seed: string) {
    this._seed = seed
    this._baseNode = HDNode.fromSeed(seed).derivePath(BASE_PATH)
    const rootNode = this._baseNode.derivePath(ROOT_STORE_PATH)
    this._keys = {
      signingKey: rootNode.derivePath('0'),
      asymEncryptionKey: nacl.box.keyPair.fromSecretKey(
        new Uint8Array(
          Buffer.from(rootNode.derivePath('2').privateKey.slice(2), 'hex'),
        ),
      ),
      symEncryptionKey: Keyring.hexToUint8Array(
        rootNode.derivePath('3').privateKey.slice(2),
      ),
      managementKey: rootNode.derivePath('1'),
    }
  }

  _getKeys() {
    return this._keys
  }

  asymEncrypt(msg, toPublic, { nonce } = {}) {
    nonce = nonce || Keyring.randomNonce()
    toPublic = naclutil.decodeBase64(toPublic)
    if (typeof msg === 'string') {
      msg = naclutil.decodeUTF8(msg)
    }
    const ephemneralKeypair = nacl.box.keyPair()
    const ciphertext = nacl.box(
      msg,
      nonce,
      toPublic,
      ephemneralKeypair.secretKey,
    )
    return {
      nonce: naclutil.encodeBase64(nonce),
      ephemeralFrom: naclutil.encodeBase64(ephemneralKeypair.publicKey),
      ciphertext: naclutil.encodeBase64(ciphertext),
    }
  }

  asymDecrypt(ciphertext, fromPublic, nonce, { space, toBuffer } = {}) {
    fromPublic = naclutil.decodeBase64(fromPublic)
    ciphertext = naclutil.decodeBase64(ciphertext)
    nonce = naclutil.decodeBase64(nonce)
    const cleartext = nacl.box.open(
      ciphertext,
      nonce,
      fromPublic,
      this._keys.asymEncryptionKey.secretKey,
    )
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  symEncrypt(msg, { space, nonce } = {}) {
    return Keyring.symEncryptBase(msg, this._keys.symEncryptionKey, nonce)
  }

  symDecrypt(ciphertext, nonce, { space, toBuffer } = {}) {
    return Keyring.symDecryptBase(
      ciphertext,
      this._keys.symEncryptionKey,
      nonce,
      toBuffer,
    )
  }

  managementPersonalSign(message) {
    const wallet = this.managementWallet()
    return wallet.signMessage(message)
  }

  managementWallet() {
    return new Wallet(this._keys.managementKey.privateKey)
  }

  getJWTSigner() {
    return SimpleSigner(this._keys.signingKey.privateKey.slice(2))
  }

  getDBSalt() {
    return sha256(this._keys.signingKey.derivePath('0').privateKey.slice(2))
  }

  getPublicKeys({ uncompressed }) {
    let signingKey = this._keys.signingKey.publicKey.slice(2)
    const managementKey = this._keys.managementKey.address
    if (uncompressed) {
      signingKey = ec
        .keyFromPublic(Buffer.from(signingKey, 'hex'))
        .getPublic(false, 'hex')
    }
    return {
      signingKey,
      managementKey,
      asymEncryptionKey: naclutil.encodeBase64(
        this._keys.asymEncryptionKey.publicKey,
      ),
    }
  }

  serialize() {
    return this._seed
  }

  static encryptWithAuthSecret(message, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symEncryptBase(message, key)
  }

  static decryptWithAuthSecret(ciphertext, nonce, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symDecryptBase(ciphertext, key, nonce)
  }

  static walletForAuthSecret(authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_WALLET,
    )
    return new Wallet(node.privateKey)
  }

  static hexToUint8Array(str) {
    return new Uint8Array(Buffer.from(str, 'hex'))
  }

  static symEncryptBase(msg, symKey, nonce) {
    nonce = nonce || Keyring.randomNonce()
    if (typeof msg === 'string') {
      msg = nacl.util.decodeUTF8(msg)
    }
    const ciphertext = nacl.secretbox(msg, nonce, symKey)
    return {
      nonce: nacl.util.encodeBase64(nonce),
      ciphertext: nacl.util.encodeBase64(ciphertext),
    }
  }

  static symDecryptBase(ciphertext, symKey, nonce, toBuffer) {
    ciphertext = nacl.util.decodeBase64(ciphertext)
    nonce = nacl.util.decodeBase64(nonce)
    const cleartext = nacl.secretbox.open(ciphertext, nonce, symKey)
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  static naclRandom(length) {
    return nacl.randomBytes(length)
  }

  static randomNonce() {
    return Keyring.naclRandom(24)
  }
}

module.exports = Keyring
