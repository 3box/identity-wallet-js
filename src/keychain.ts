import type { ThreeIDX, AuthEntry, NewAuthEntry, EncData } from './three-idx'
import type { DidProvider } from './did-provider'
import Keyring, { LATEST } from './keyring'
import { asymDecryptJWE, asymEncryptJWE, parseJWEKids } from './crypto'
import { encodeKey, decodeKey, fakeEthProvider } from './utils'

import { createLink } from '3id-blockchain-utils'
import { AccountID } from 'caip'

async function decryptAuthId(encrypted: EncData, keyring: Keyring): Promise<string | null> {
  if (!encrypted.jwe) throw new Error('Invalid encrypted block')
  const decrypter = keyring.getAsymDecrypter(parseJWEKids(encrypted.jwe))
  const decrypted = await asymDecryptJWE(encrypted.jwe, { decrypter })
  return decrypted.id as string
}

export async function newAuthEntry(
  keyring: Keyring,
  did: string,
  authId: string,
  authSecret: Uint8Array
): Promise<NewAuthEntry> {
  const mainPub = keyring.getEncryptionPublicKey()
  const mainKid = `${did}#${keyring.getKeyFragment(LATEST, true)}`
  const { publicKey } = Keyring.authSecretToKeyPair(authSecret)
  const wallet = Keyring.authSecretToWallet(authSecret)
  const accountId = new AccountID({ address: wallet.address, chainId: 'eip155:1' })
  const cleartext: Record<string, any> = { seed: keyring.seed }
  // If we have a legacy seed v03ID will be defined
  if (keyring.v03ID) cleartext.v03ID = keyring.v03ID
  return {
    pub: encodeKey(publicKey, 'x25519'),
    data: { jwe: await asymEncryptJWE(cleartext, { publicKey }) },
    id: { jwe: await asymEncryptJWE({ id: authId }, { publicKey: mainPub, kid: mainKid }) },
    linkProof: await createLink(did, accountId, fakeEthProvider(wallet)),
  }
}

export async function updateAuthEntry(
  keyring: Keyring,
  authEntry: AuthEntry,
  removedAuthIds: Array<string>,
  did: string
): Promise<AuthEntry | null> {
  const mainPub = keyring.getEncryptionPublicKey()
  const mainKid = `${did}#${keyring.getKeyFragment(LATEST, true)}`
  const publicKey = decodeKey(authEntry.pub)
  const authId = await decryptAuthId(authEntry.id, keyring)
  // Return null if auth entry should be removed
  if (removedAuthIds.find((id) => id === authId)) return null
  return Object.assign(authEntry, {
    data: { jwe: await asymEncryptJWE({ seed: keyring.seed }, { publicKey }) },
    id: { jwe: await asymEncryptJWE({ id: authId }, { publicKey: mainPub, kid: mainKid }) },
  })
}

async function rotateKeys(
  threeIdx: ThreeIDX,
  keyring: Keyring,
  removedAuthIds: Array<string>
): Promise<void> {
  const version = await threeIdx.get3idVersion()
  await keyring.generateNewKeys(version)
  const update3idState = keyring.get3idState()
  const pastSeeds = keyring.pastSeeds
  const updatedAuthEntries = (
    await Promise.all(
      threeIdx
        .getAllAuthEntries()
        .map((entry) => updateAuthEntry(keyring, entry, removedAuthIds, threeIdx.id))
    )
  ).filter((entry) => Boolean(entry)) as Array<AuthEntry> // filter removes null entires
  await threeIdx.rotateKeys(update3idState, pastSeeds, updatedAuthEntries)
}

interface PendingAdd {
  authId: string
  entry: NewAuthEntry
}

interface KeychainStatus {
  clean: boolean
  adding: Array<string>
  removing: Array<string>
}

export class Keychain {
  private _pendingAdds: Array<PendingAdd> = []
  private _pendingRms: Array<string> = []
  /**
   * The Keychain enables adding and removing of authentication methods.
   */
  constructor(public _keyring: Keyring, protected _threeIdx: ThreeIDX) {}

  /**
   * List all current authentication methods.
   *
   * @return    {Array<string>}                           A list of authIds.
   */
  async list(): Promise<Array<string>> {
    return Promise.all(
      this._threeIdx.getAllAuthEntries().map(
        async ({ id }: AuthEntry): Promise<string> => {
          return (await decryptAuthId(id, this._keyring)) as string
        }
      )
    )
  }

  /**
   * Add a new authentication method (adds to staging).
   *
   * @param     {String}            authId          An identifier for the auth method
   * @param     {Uint8Array}        authSecret      The authSecret to use, should be of sufficient entropy
   */
  async add(authId: string, authSecret: Uint8Array): Promise<void> {
    this._pendingAdds.push({
      authId,
      entry: await newAuthEntry(this._keyring, this._threeIdx.id, authId, authSecret),
    })
  }

  /**
   * Remove an authentication method (adds to staging).
   *
   * @param     {String}            authId          An identifier for the auth method
   */
  async remove(authId: string): Promise<void> { // eslint-disable-line
    this._pendingRms.push(authId)
  }

  /**
   * Show the staging status of the keychain.
   * Since removing auth methods will rotate the keys of the 3ID its a good idea
   * to remove multiple auth methods at once if desired. Therefore we introduce
   * a commit pattern to do multiple updates to the keychain at once.
   *
   * @return    {KeychainStatus}                    Object that states the staging status of the keychain
   */
  status(): KeychainStatus {
    return {
      clean: !(this._pendingAdds.length + this._pendingRms.length),
      adding: this._pendingAdds.map((e) => e.authId),
      removing: this._pendingRms,
    }
  }

  /**
   * Commit the staged changes to the keychain.
   */
  async commit(): Promise<void> {
    if (!this._pendingAdds.length && !this._pendingRms.length) throw new Error('Nothing to commit')
    if (this._threeIdx.getAllAuthEntries().length === 0) {
      if (this._pendingRms.length) throw new Error('Can not remove non-existent auth method')
      if (!this._pendingAdds.length) throw new Error('Can not add non-existent auth method')
      // Create IDX structure if not present
      await this._threeIdx.createIDX(this._pendingAdds.pop()?.entry)
    }
    if (this._pendingRms.length) {
      await rotateKeys(this._threeIdx, this._keyring, this._pendingRms)
      this._pendingRms = []
    }
    if (this._pendingAdds.length) {
      const entries = this._pendingAdds.map((e) => e.entry)
      this._pendingAdds = []
      await this._threeIdx.addAuthEntries(entries)
    }
  }

  static async load(
    threeIdx: ThreeIDX,
    authSecret: Uint8Array,
    makeTmpProvider: (keyring: Keyring, managementKey: string) => DidProvider
  ): Promise<Keychain> {
    const { secretKey } = Keyring.authSecretToKeyPair(authSecret)
    const wallet = Keyring.authSecretToWallet(authSecret)
    const accountId = new AccountID({ address: wallet.address.toLowerCase(), chainId: 'eip155:1' })
    const authData = await threeIdx.loadIDX(accountId.toString())
    if (authData) {
      if (!authData.seed?.jwe) throw new Error('Unable to find auth data')
      try {
        const decrypted = await asymDecryptJWE(authData.seed.jwe, { secretKey })
        // If we have a legacy seed v03ID will be defined
        const keyring = new Keyring(new Uint8Array(decrypted.seed), decrypted.v03ID)
        await keyring.loadPastSeeds(authData.pastSeeds)
        // We might have the v03ID in the past seeds, check and set if present
        if (keyring.v03ID) threeIdx.setV03ID(keyring.v03ID)
        return new Keychain(keyring, threeIdx)
      } catch (e) {
        if (e.message === 'Failed to decrypt') throw new Error('Auth not allowed')
        throw e
      }
    }
    return Keychain.create(threeIdx, makeTmpProvider)
  }

  static async create(
    threeIdx: ThreeIDX,
    makeTmpProvider: (keyring: Keyring, managementKey: string) => DidProvider,
    seed?: Uint8Array,
    v03ID?: string
  ): Promise<Keychain> {
    const keyring = new Keyring(seed, v03ID)
    const docParams = keyring.get3idState(true)
    // Temporarily set DID provider to create 3ID document
    await threeIdx.setDIDProvider(makeTmpProvider(keyring, docParams.metadata.owners[0]))
    await threeIdx.create3idDoc(docParams)
    if (v03ID) threeIdx.setV03ID(v03ID)
    return new Keychain(keyring, threeIdx)
  }
}
