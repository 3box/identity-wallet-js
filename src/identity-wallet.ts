import { CeramicApi } from '@ceramicnetwork/ceramic-common'

import { DidProvider } from './did-provider'
import Keyring from './keyring'
import ThreeIdProvider from './threeIdProvider'
import { ThreeIDX } from './three-idx'
import Permissions, { GetPermissionFn, SELF_ORIGIN } from './permissions'
import { Keychain } from './keychain'

interface IDWConfig {
  getPermission: GetPermissionFn
  seed?: string
  authSecret?: Uint8Array
  authId?: string
  externalAuth?: (req: any) => Promise<any>
  ceramic: CeramicApi
  useThreeIdProv: boolean
}

export default class IdentityWallet {
  protected _externalAuth: ((req: any) => Promise<any>) | undefined

  /**
   * Use IdentityWallet.create() to create an IdentityWallet instance
   */
  constructor(
    protected _keyring: Keyring,
    protected _threeIdx: ThreeIDX,
    protected _permissions: Permissions,
    protected _keychain: Keychain
  ) {}
  /**
   * @property {Keychain} keychain          Edit the keychain
   */
  get keychain(): Keychain {
    return this._keychain
  }

  /**
   * @property {Permissions} permissions    Edit permissions
   */
  get permissions(): Permissions {
    return this._permissions
  }

  /**
   * @property {string} DID                 The 3ID of the IdentityWallet instance
   */
  get DID(): string {
    return this._threeIdx.DID
  }

  /**
   * Creates an instance of IdentityWallet
   *
   * @param     {Object}        config                  The configuration to be used
   * @param     {Function}      config.getPermission    The function that is called to ask the user for permission
   * @param     {String}        config.seed             The seed of the identity, 32 bytes hex string
   * @param     {Uint8Array}    config.authSecret       The authSecret to use, 32 bytes
   * @param     {String}        config.authId           The authId is used to identify the authSecret
   * @param     {String}        config.externalAuth     External auth function, directly returns key material, used to migrate legacy 3box accounts
   * @return    {IdentityWallet}                        An IdentityWallet instance
   */
  static async create(config: IDWConfig): Promise<IdentityWallet> {
    if (config.seed && config.authSecret) throw new Error("Can't use both seed and authSecret")
    if (!config.seed && !config.authSecret) throw new Error('Either seed or authSecret is needed')
    if (config.authSecret && !config.authId) {
      throw new Error('AuthId must be given along with authSecret')
    }
    const threeIdx = new ThreeIDX(config.ceramic)
    const permissions = new Permissions(config.getPermission)
    let keyring, keychain
    if (config.seed) {
      keyring = new Keyring(config.seed)
      keychain = new Keychain(keyring, threeIdx)
      const pubkeys = keyring.getPublicKeys({ mgmtPub: true, useMulticodec: true })
      await threeIdx.create3idDoc(pubkeys)
    } else if (config.authSecret) {
      keychain = await Keychain.load(threeIdx, config.authSecret)
      keyring = keychain._keyring
    }
    permissions.setDID(threeIdx.DID)
    const idw = new IdentityWallet(keyring as Keyring, threeIdx, permissions, keychain as Keychain)
    await idw._threeIdx.setDIDProvider(idw.getDidProvider(SELF_ORIGIN))
    if (config.authId && !keychain?.list().length) {
      // Add the auth method to the keychain
      await idw.keychain.add(config.authId, config.authSecret as Uint8Array)
      await idw.keychain.commit()
    }
    return idw
  }

  /**
   * Get the DIDProvider
   *
   * @return    {DidProvider}                   The DIDProvider for this IdentityWallet instance
   */
  getDidProvider(forcedOrigin?: string): DidProvider {
    return new DidProvider({
      keyring: this._keyring,
      permissions: this.permissions,
      threeIdx: this._threeIdx,
      forcedOrigin,
    })
  }

  /**
   * Get the 3IDProvider
   *
   * @return    {ThreeIdProvider}                   The 3IDProvider for this IdentityWallet instance
   */
  get3idProvider(forcedOrigin?: string): ThreeIdProvider {
    return new ThreeIdProvider({
      keyring: this._keyring,
      permissions: this.permissions,
      threeIdx: this._threeIdx,
      forcedOrigin,
    })
  }
}
