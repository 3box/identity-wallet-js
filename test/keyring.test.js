import { mnemonicToSeed } from '@ethersproject/hdnode'
import * as u8a from 'uint8arrays'

import { randomBytes } from '../src/crypto'
import Keyring from '../src/keyring'

describe('Keyring', () => {
  const seed = u8a.fromString('f0e4c2f76c58916ec258f246851bea091d14d4247a2fc3e18694461b1816e13b', 'base16')

  it('Generates random seed if none passed', async () => {
    const keyring = new Keyring()
    expect(keyring.seed).toBeDefined()
  })

  it('Derives correct keys from seed', async () => {
    const keyring = new Keyring(seed)
    expect(keyring.seed).toEqual(seed)
    expect(keyring.pastSeeds).toEqual([])
    expect(keyring.get3idState(true)).toMatchSnapshot()
    expect(keyring.get3idState()).toMatchSnapshot()
    expect(keyring.getEncryptionPublicKey()).toMatchSnapshot()
    expect(keyring.seed).toMatchSnapshot()
  })

  it('generate and load keys', async () => {
    const seed = randomBytes(32)
    const keyring = new Keyring(seed)
    expect(keyring.seed).toEqual(seed)
    const signed0 = await keyring.getSigner()('asdf')
    const mgmt0 = keyring.get3idState().metadata.owners[0].split(':')[2]

    const v0 = 'versionCID0'
    await keyring.generateNewKeys(v0)
    expect(keyring.seed).not.toEqual(seed)
    expect(await keyring.getSigner(v0)('asdf')).toEqual(signed0)
    const signed1 = await keyring.getSigner()('asdf')
    expect(signed1).not.toEqual(signed0)
    const mgmt1 = keyring.get3idState().metadata.owners[0].split(':')[2]

    const v1 = 'versionCID1'
    await keyring.generateNewKeys(v1)
    expect(keyring.seed).not.toEqual(seed)
    expect(await keyring.getSigner(v0)('asdf')).toEqual(signed0)
    expect(await keyring.getSigner(v1)('asdf')).toEqual(signed1)
    const signed2 = await keyring.getSigner()('asdf')
    expect(signed2).not.toEqual(signed0)
    expect(signed2).not.toEqual(signed1)
    const docState2 = keyring.get3idState()
    const mgmt2 = keyring.get3idState().metadata.owners[0].split(':')[2]

    const mgmtSigs = await Promise.all([
      keyring.getMgmtSigner(mgmt0)('asdf'),
      keyring.getMgmtSigner(mgmt1)('asdf'),
      keyring.getMgmtSigner(mgmt2)('asdf')
    ])

    const pastSeeds = keyring.pastSeeds
    const keyring1 = new Keyring(keyring.seed)
    expect(await keyring1.getSigner()('asdf')).toEqual(signed2)
    await keyring1.loadPastSeeds(pastSeeds)
    expect(await keyring1.getSigner(v0)('asdf')).toEqual(signed0)
    expect(await keyring1.getSigner(v1)('asdf')).toEqual(signed1)
    expect(keyring1.get3idState()).toEqual(docState2)

    expect(await Promise.all([
      keyring1.getMgmtSigner(mgmt0)('asdf'),
      keyring1.getMgmtSigner(mgmt1)('asdf'),
      keyring1.getMgmtSigner(mgmt2)('asdf')
    ])).toEqual(mgmtSigs)
  })

  it('generateNewKeys throws if version already exist', async () => {
    const keyring = new Keyring()
    const v = 'versionCID0'
    await keyring.generateNewKeys(v)
    expect(keyring.generateNewKeys(v)).rejects.toThrow('Key set version already exist')
  })
})
