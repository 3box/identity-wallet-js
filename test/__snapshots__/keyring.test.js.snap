// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyring Derives correct keys from seed 1`] = `
Object {
  "content": Object {
    "publicKeys": Object {
      "PpEHx9GhyFT5Env": "zQ3shnyMbHDVqqJnm6EiPJ47Vc4ivUFgcaPpEHx9GhyFT5Env",
      "suxjL1MxhcRFYkG": "z6DtSycrUdhF2Cv9vQP9GoTVsbjUsSuj5suxjL1MxhcRFYkG",
    },
  },
  "metadata": Object {
    "isUnique": false,
    "owners": Array [
      "did:key:zQ3shjdQdzyD4TLGhSrGZenUjS3TnhwqWowkqKqi2J8sMgGy6",
    ],
    "tags": Array [
      "3id",
    ],
  },
}
`;

exports[`Keyring Derives correct keys from seed 2`] = `
Object {
  "content": Object {
    "publicKeys": Object {
      "PpEHx9GhyFT5Env": "zQ3shnyMbHDVqqJnm6EiPJ47Vc4ivUFgcaPpEHx9GhyFT5Env",
      "suxjL1MxhcRFYkG": "z6DtSycrUdhF2Cv9vQP9GoTVsbjUsSuj5suxjL1MxhcRFYkG",
    },
  },
  "metadata": Object {
    "owners": Array [
      "did:key:zQ3shjdQdzyD4TLGhSrGZenUjS3TnhwqWowkqKqi2J8sMgGy6",
    ],
  },
}
`;

exports[`Keyring Derives correct keys from seed 3`] = `
Uint8Array [
  77,
  18,
  225,
  143,
  155,
  176,
  47,
  185,
  247,
  228,
  117,
  64,
  200,
  168,
  40,
  10,
  160,
  46,
  113,
  26,
  139,
  93,
  200,
  71,
  107,
  247,
  97,
  122,
  195,
  166,
  57,
  57,
]
`;

exports[`Keyring Derives correct keys from seed 4`] = `
Uint8Array [
  240,
  228,
  194,
  247,
  108,
  88,
  145,
  110,
  194,
  88,
  242,
  70,
  133,
  27,
  234,
  9,
  29,
  20,
  212,
  36,
  122,
  47,
  195,
  225,
  134,
  148,
  70,
  27,
  24,
  22,
  225,
  59,
]
`;
